accounts/abi/bind/dep_tree.go

```diff
@@ -0,0 +1,234 @@
+type depTreeDeployer struct {
+	deployedAddrs map[string]common.Address
+	deployerTxs   map[string]*types.Transaction
+	input         map[string][]byte // map of the root contract pattern to the constructor input (if there is any)
+	deploy        DeployFn
+}
+
+// linkAndDeploy recursively deploys a contract and its dependencies:  starting by linking/deploying its dependencies.
+// The deployment result (deploy addresses/txs or an error) is stored in the depTreeDeployer object.
+func (d *depTreeDeployer) linkAndDeploy(node *depTreeNode) error {
+	// don't deploy contracts specified as overrides.  don't deploy their dependencies.
+	if node.overrideAddr != nil {
+		return nil
+	}
+
+	// if this contract/library depends on other libraries deploy them (and their dependencies) first
+	for _, childNode := range node.children {
+		if err := d.linkAndDeploy(childNode); err != nil {
+			return err
+		}
+	}
+	// if we just deployed any prerequisite contracts, link their deployed addresses into the bytecode to produce
+	// a deployer bytecode for this contract.
+	deployerCode := node.unlinkedCode
+	for _, child := range node.children {
+		linkAddr := d.deployedAddrs[child.pattern]
+		if child.overrideAddr != nil {
+			linkAddr = *child.overrideAddr
+		}
+		deployerCode = strings.ReplaceAll(deployerCode, "__$"+child.pattern+"$__", strings.ToLower(linkAddr.String()[2:]))
+	}
+
+	// Finally, deploy the contract.
+	addr, tx, err := d.deploy(d.input[node.pattern], common.Hex2Bytes(deployerCode))
+	if err != nil {
+		return err
+	}
+
+	d.deployedAddrs[node.pattern] = addr
+	d.deployerTxs[node.pattern] = tx
+	return nil
+}
+
+// result returns a result for this deployment, or an error if it failed.
+func (d *depTreeDeployer) result() *DeploymentResult {
+	return &DeploymentResult{
+		Txs:   d.deployerTxs,
+		Addrs: d.deployedAddrs,
+	}
+}
+
+func newDepTreeDeployer(deploy DeployFn) *depTreeDeployer {
+	return &depTreeDeployer{
+		deploy:        deploy,
+		deployedAddrs: make(map[string]common.Address),
+		deployerTxs:   make(map[string]*types.Transaction)}
+}
+
+// LinkAndDeploy deploys a specified set of contracts and their dependent
+// libraries.  If an error occurs, only contracts which were successfully
+// deployed are returned in the result.
+func LinkAndDeploy(deployParams *DeploymentParams, deploy DeployFn) (res *DeploymentResult, err error) {
+	unlinkedContracts := make(map[string]string)
+	accumRes := &DeploymentResult{
+		Txs:   make(map[string]*types.Transaction),
+		Addrs: make(map[string]common.Address),
+	}
+	for _, meta := range deployParams.contracts {
+		unlinkedContracts[meta.Pattern] = meta.Bin[2:]
+	}
+	treeBuilder := newDepTreeBuilder(deployParams.overrides, unlinkedContracts)
+	deps, _ := treeBuilder.BuildDepTrees()
+
+	for _, tr := range deps {
+		deployer := newDepTreeDeployer(deploy)
+		if deployParams.inputs != nil {
+			deployer.input = map[string][]byte{tr.pattern: deployParams.inputs[tr.pattern]}
```

Comment:
Why is it named `tr`? It's definitely not obvious to a casual reader. 
And what does `pattern` signify, really? Please document or name it more appropriate. For me, it's a bit black magic what is happening here, with the setting of `deployer.input` to a map, involving the pattern, if the `deployParams.inputs` is non-nil.

Reply:
`tr` represents the root of a deployment tree here: it's a contract/library that isn't a dependency of any other contracts.

a contract's pattern is a legitimate solidity concept ([reference](https://docs.soliditylang.org/en/latest/using-the-compiler.html#library-linking)): it's the "34 character prefix of the hex encoding of the keccak256 hash of the fully qualified library name".  A unique identifier for each contract/library.

I think for clarity, it probably deserves its own type (even if just a typedef on string).